### PR TITLE
Added HtmlPurifier to improve MaliciousCode filtering

### PR DESCRIPTION
**Pull request by Judx**
### Description (*)

Add HtmlPurifier for improved MaliciousCode filtering. Found when researching https://helpx.adobe.com/security/products/magento/apsb23-50.html

[Magento Open Source > 2.4.5](https://experienceleague.adobe.com/docs/commerce-operations/release/notes/magento-open-source/2-4-5.html?lang=en) added HtmlPurifier library to improve the MaliciousCode filter.
https://github.com/OpenMage/magento-lts/discussions/3600

### Related Pull Requests

https://github.com/OpenMage/magento-lts/security/advisories/GHSA-3p73-mm7v-4f6m

### Fixed Issues (if relevant)

1. https://github.com/OpenMage/magento-lts/discussions/3600

### Manual testing scenarios (*)

1. Call `Mage_Core_Model_Input_Filter_MaliciousCode::filter()` with [dirty html](http://htmlpurifier.org/live/smoketests/xssAttacks.php)
2. The value is filtered with HTMLPurifier.

### Questions or comments

Could someone run `n98-magerun.phar dev:ide:phpstorm:meta` for the new helper meta?

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [X] Add yourself to contributors list

### CODE REVIEW

File: app/code/core/Mage/Core/Helper/Purifier.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * OpenMage
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available at https://opensource.org/license/osl-3-0-php
+ *
+ * @category   Mage
+ * @package    Mage_Core
+ * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
+ * @copyright  Copyright (c) 2019-2023 The OpenMage Contributors (https://www.openmage.org)
```
Comment (fballiano) [inline app/code/core/Mage/Core/Helper/Purifier.php:12]:
```suggestion
 * @copyright  Copyright (c) 2023 The OpenMage Contributors (https://www.openmage.org)
```

File: app/code/core/Mage/Core/Helper/Purifier.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     *
+     * @param array|string $content
+     * @return mixed
+     */
+    public function purify($content): mixed
+    {
+        return is_array($content) ? $this->purifier->purifyArray($content) : $this->purifier->purify($content);
+    }
+
+}
```
Comment (fballiano) [inline app/code/core/Mage/Core/Helper/Purifier.php:54]:
```suggestion
}

```

File: app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php [hunk 3]
```diff
@@ -56,10 +56,10 @@ class Mage_Core_Model_Input_Filter_MaliciousCode implements Zend_Filter_Interfac
     public function filter($value)
     {
         do {
-            $value = preg_replace($this->_expressions, '', $value, -1, $count);
+            $value = preg_replace($this->_expressions, '', $value ?? '', -1, $count);
         } while ($count !== 0);
 
-        return $value;
+        return Mage::helper('purifier')->purify($value);
```
Comment (ADDISON74) [inline app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php:62]: PHPStan is a little bit upset here.

Comment (sreichel) [reply app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php:62]: `Mage::helper('core/purifier')` ?

Comment (pquerner) [reply app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php:62]: Probably not using helper variable.

**Comment by Judx** [comment]
Thank you for your contribution. I have improved the PR as suggested.

File: app/code/core/Mage/Core/Helper/Purifier.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * OpenMage
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available at https://opensource.org/license/osl-3-0-php
+ *
+ * @category   Mage
+ * @package    Mage_Core
+ * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
+ * @copyright  Copyright (c) 2019-2023 The OpenMage Contributors (https://www.openmage.org)
```
Comment (fballiano) [inline app/code/core/Mage/Core/Helper/Purifier.php:12]:
```suggestion
 * @copyright  Copyright (c) 2023 The OpenMage Contributors (https://www.openmage.org)
```

because this is a new file if i'm not mistaken

File: app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php [hunk 5]
```diff
@@ -8,8 +8,7 @@
  *
  * @category   Mage
  * @package    Mage_Core
- * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
- * @copyright  Copyright (c) 2019-2023 The OpenMage Contributors (https://www.openmage.org)
+ * @copyright  Copyright (c) 2023 The OpenMage Contributors (https://www.openmage.org)
```
Comment (fballiano) [inline app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php:11]:
```suggestion
* @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
 * @copyright  Copyright (c) 2019-2023 The OpenMage Contributors (https://www.openmage.org)
```

while this file already exists so we've to keep the previous copyright (maybe I suggested a wrong change?)

**Comment by fballiano** [comment]
seems like this version of htmlpurify is php8.0+ so we can't use it in `main`, so:
- we could downgrade htmlpurify
- add this PR to `next`

**Comment by Judx** [comment]
Good point. Which one do you suggest? I am down to downgrading htmlpurify as its mostly syntax changes php7 -> php8

**Comment by fballiano** [comment]
well, I'm biased towards releasing the `next` branch so... if you ask me... but maybe v4.15 could be added to `main` (or the latest php7 compatible version, I couldn't really find that info) and the latest one to `next`

**Comment by ADDISON74** [comment]
There are two errors reported by PHPStans. the latest changes are not compatible with lower versions of PHP8. I would leave it as it was initially so as not to create a BC of all the beauty.

It is a good an important addition that I would leave in the main branch.

File: app/code/core/Mage/Core/Helper/Purifier.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        $this->purifier = $purifier ?? new HTMLPurifier($config);
+    }
+
+    /**
+     * Purify Html Content
+     *
+     * @param array|string $content
+     * @return array|string
+     */
+    public function purify(array|string $content): array|string
```
Comment (fballiano) [inline app/code/core/Mage/Core/Helper/Purifier.php:48]: ahhh the PHP8 problem is just about this line, not an incompatibility with htmlpurify itself, so it should be easier to solve

**Comment by Judx** [comment]
Everything seems to be correct now? Is it possible to squash commits while merging or should I do that?

**Comment by fballiano** [comment]
do you know how to trigger the filter() method in order to test this?

**Comment by Judx** [comment]
> do you know how to trigger the filter() method in order to test this?

If you have n98-magerun try this:
1. `n98-magerun dev:console`
2. `Mage::getModel('core/Input_Filter_MaliciousCode')->filter('test')` -> `test`
3. `Mage::getModel('core/Input_Filter_MaliciousCode')->filter("'';!--\"<XSS>=&{()}")` -> `'';!--"=&amp;{()}`

Different attacks and correct results can be found here: http://htmlpurifier.org/live/smoketests/xssAttacks.php

**Comment by Judx** [comment]
We might want to publish security advisory for this too if we want people to update fast.

**Comment by fballiano** [comment]
thanks @Judx, I was wondering where (in the adminhtml) is called that filter method, but anyway I like this PR a lot

**Comment by Judx** [comment]
> thanks @Judx, I was wondering where (in the adminhtml) is called that filter method, but anyway I like this PR a lot

Great! I can find total 7 usages, in adminhtml i.e. `Mage_Adminhtml_Block_Template`:
```
        magento-lts  (7 usages found)
            app/code/core/Mage/Adminhtml/Block  (1 usage found)
                Template.php  (1 usage found)
                    Mage_Adminhtml_Block_Template  (1 usage found)
                        maliciousCodeFilter  (1 usage found)
                            78 return Mage::getSingleton('core/input_filter_maliciousCode')->filter($html);
            app/code/core/Mage/Adminhtml/controllers  (1 usage found)
                AjaxController.php  (1 usage found)
                    Mage_Adminhtml_AjaxController  (1 usage found)
                        translateAction  (1 usage found)
                            37 $item['custom'] = $filter->filter($item['custom']);
            app/code/core/Mage/Cms/Block/Widget  (1 usage found)
                Block.php  (1 usage found)
                    Mage_Cms_Block_Widget_Block  (1 usage found)
                        _beforeToHtml  (1 usage found)
                            73 Mage::getSingleton('core/input_filter_maliciousCode')->filter($block->getContent())
            app/code/core/Mage/Core/controllers  (1 usage found)
                AjaxController.php  (1 usage found)
                    Mage_Core_AjaxController  (1 usage found)
                        translateAction  (1 usage found)
                            36 $item['custom'] = $filter->filter($item['custom']);
            app/code/core/Mage/Core/Model/Input  (1 usage found)
                Filter.php  (1 usage found)
                    Mage_Core_Model_Input_Filter  (1 usage found)
                        _filter  (1 usage found)
                            230 $value = $zendFilter->filter($value);
            app/code/core/Mage/Customer/Block/Address/Renderer  (1 usage found)
                Default.php  (1 usage found)
                    Mage_Customer_Block_Address_Renderer_Default  (1 usage found)
                        _prepareAddressTemplateData  (1 usage found)
                            149 $result = preg_replace($urlRegExp, ' ', $maliciousCodeFilter->filter($data));
            app/code/core/Mage/ProductAlert/Block/Email  (1 usage found)
                Abstract.php  (1 usage found)
                    Mage_ProductAlert_Block_Email_Abstract  (1 usage found)
                        _getFilteredProductShortDescription  (1 usage found)
                            136 $shortDescription = Mage::getSingleton('core/input_filter_maliciousCode')->filter($shortDescription);
```

**Comment by sreichel** [comment]
Think doc block `array|string` was better, or do i miss something?

**Comment by mattdavenport** [comment]
If there is any interest, we have some functionality added on top of HtmlPurifier internally which enhances security by preventing javascript tags from being inserted into areas such as system config, categories, CMS, etc. If there is any interest I could prepare a branched PR (or separate issue). I do assume there are some legitimate use cases for JS in these areas though, so if there is any interest we could put it behind a configuration flag (e.g. "Allow JS in text inputs").

**Comment by Judx** [comment]
Waiting for merge! And superthanks to @mattdavenport would be great if you could improve this too! :)

File: app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php [hunk 7]
```diff
@@ -26,7 +26,7 @@ class Mage_Core_Model_Input_Filter_MaliciousCode implements Zend_Filter_Interfac
      *
      * @var array
      */
-    protected $_expressions = [
+    protected array $_expressions = [
```
Comment (sreichel) [inline app/code/core/Mage/Core/Model/Input/Filter/MaliciousCode.php:29]: Adding types should be done in another PR.

**Comment by sreichel** [comment]
lgtm

**Comment by mattdavenport** [comment]
@Judx Apologies for the delay here. After further inspection it appears that our improvements have been made redundant by later versions of HTMLPurifier. Nonetheless, thank you for the this PR!

File: composer.json [hunk 8]
```diff
@@ -25,6 +25,7 @@
         "colinmollenhour/cache-backend-redis": "^1.14",
         "colinmollenhour/magento-redis-session": "^3.2.0",
         "cweagans/composer-patches": "^1.7",
+        "ezyang/htmlpurifier": "^4.16",
```
Comment (fballiano) [inline composer.json:28]:
```suggestion
        "ezyang/htmlpurifier": "^4.17",
```

**Review by fballiano — APPROVED** [review]

File: .phpstorm.meta.php/magento_helpers.meta.php [hunk 9]
```diff
@@ -24,10 +24,10 @@
             'adminhtml/media_js' => \Mage_Adminhtml_Helper_Media_Js::class,
             'adminhtml/rss' => \Mage_Adminhtml_Helper_Rss::class,
             'adminhtml/sales' => \Mage_Adminhtml_Helper_Sales::class,
-            'api2/data' => \Mage_Api2_Helper_Data::class,
             'api2' => \Mage_Api2_Helper_Data::class,
-            'api/data' => \Mage_Api_Helper_Data::class,
+            'api2/data' => \Mage_Api2_Helper_Data::class,
```
Comment (kiatng) [inline .phpstorm.meta.php/magento_helpers.meta.php:28]: I am not sure how is this related to the PR.

Comment (fballiano) [reply .phpstorm.meta.php/magento_helpers.meta.php:28]: I think @Judx ran n98-magerun to generate the phpstorm helper because of the new classed that are added including the new module thru composer. and I seem to recall that the order is messed up depending on the php version that's been used, but that same line is just a few lines above so it seems correct.

**Review by fballiano — APPROVED** [review]

**Review by ADDISON74 — APPROVED** [review]
